### PR TITLE
Update doc to state that CookieTempData is the default

### PR DESCRIPTION
**Pull request by Eilon**
In ASP.NET Core 2.0 the Cookie TempData provider is the default, not session state.

Fixes https://github.com/aspnet/Docs/issues/3767

[review URL](https://review.docs.microsoft.com/en-us/aspnet/core/fundamentals/app-state?branch=pr-en-us-4350)

### CODE REVIEW

**Comment by Eilon** [comment]
Here's the current doc: https://docs.microsoft.com/en-us/aspnet/core/fundamentals/app-state?tabs=aspnetcore2x

**Comment by Eilon** [comment]
cc @marlon-tucker

File: aspnetcore/fundamentals/app-state.md [hunk 1]
```diff
@@ -42,15 +42,15 @@ ASP.NET Core MVC exposes the [TempData](https://docs.microsoft.com/aspnet/core/a
 
 ## Cookie-based TempData provider 
 
-In ASP.NET Core 1.1 and higher, you can use the cookie-based TempData provider to store a user's TempData in a cookie. To enable the  cookie-based TempData provider, register the `CookieTempDataProvider` service in `ConfigureServices`:
+In ASP.NET Core 2.0 and higher, the cookie-based TempData provider is used by default to store a user's TempData in a cookie. In ASP.NET Core 1.0 and 1.1 the session state TempData provider is the default. To configure the TempData provider for an application, register a TempData provider implementation in `ConfigureServices`:
```
Comment (guardrex) [inline aspnetcore/fundamentals/app-state.md]: Add a comma ...

> In ASP.NET Core 1.0 and 1.1, the session state ...

**Comment by marlon-tucker** [comment]
Hello

Is `services.AddSingleton...` the recommended way to do this or is `service.AddMvc().AddSessionStateTempDataProvider();`

The later is what the origial issue suggested, and as this doc is pretty high up on google results for state in ASP Core it's good to nudge people in the right direction.

I realise they achieve effectively the same thing.

**Comment by Eilon** [comment]
I'll work on the feedback, thanks @marlon-tucker and @guardrex for reviewing!

**Comment by guardrex** [comment]
@Eilon This covers https://github.com/aspnet/Docs/issues/3767?

**Comment by Eilon** [comment]
@guardrex yes it does! I tried searching for bugs but didn't find that one. I searched for `TempData` and not `Temp Data` so I didn't find it.

**Comment by Eilon** [comment]
cc @rynowak who logged https://github.com/aspnet/Docs/issues/3767.

File: aspnetcore/fundamentals/app-state.md [hunk 2]
```diff
@@ -42,20 +42,26 @@ ASP.NET Core MVC exposes the [TempData](https://docs.microsoft.com/aspnet/core/a
...
 
-In ASP.NET Core 2.0 and higher, the cookie-based TempData provider is used by default to store a user's TempData in a cookie. In ASP.NET Core 1.0 and 1.1 the session state TempData provider is the default. To configure the TempData provider for an application, register a TempData provider implementation in `ConfigureServices`:
+In ASP.NET Core 2.0 and higher, the cookie-based TempData provider is used by default to store a user's TempData in a cookie.
+
+The cookie data is encoded with the [Base64UrlTextEncoder](https://docs.microsoft.com/aspnet/core/api/microsoft.aspnetcore.authentication.base64urltextencoder). Because the cookie is encrypted and chunked, the single cookie size limit does not apply. The cookie data is not compressed, because compressing encryped data can lead to security problems such as the [CRIME](https://wikipedia.org/wiki/CRIME_(security_exploit)) and [BREACH](https://wikipedia.org/wiki/BREACH_(security_exploit)) attacks. For more information on the cookie-based TempData provider, see [CookieTempDataProvider](https://github.com/aspnet/Mvc/blob/dev/src/Microsoft.AspNetCore.Mvc.ViewFeatures/ViewFeatures/CookieTempDataProvider.cs).
```
Comment (guardrex) [inline aspnetcore/fundamentals/app-state.md]: Use a relative link there ...

> \[Base64UrlTextEncoder](/aspnet/core/api/microsoft.aspnetcore.authentication.base64urltextencoder)

Use the contraction: "is not" :point_right: "isn't"

Drop the comma before "because" since the statement that follows is restrictive:

> ... is not compressed because compressing ...

Comment (Eilon) [reply aspnetcore/fundamentals/app-state.md]: BTW I didn't change that text at all, I just moved where the paragraph is. Do you feel these improvements are important?

Comment (guardrex) [reply aspnetcore/fundamentals/app-state.md]: lol I thought you typed that. :smile:

No, that's ok. I'll take care of it. I'll do a quick pass on this topic and clean it up later.

**Comment by rynowak** [comment]
This needs some explanation about the tradeoffs and what we recommend. 

For instance you should use session if you need to store a lot of data or if you rely on temp data for application correctness. If you are using sessions in general then you probably want to choose it for temp data.

**Comment by guardrex** [comment]
There's also a tracking issue for updates to the app-state topic: https://github.com/aspnet/Docs/issues/3216

@scottaddie @rynowak I could jump on #3216 after this PR goes in. The topic needs a little guidelines adherence and 1.x sample help anyway. I can get your notes in (along with anything else u want) on that update.

**Comment by Rick-Anderson** [comment]
@rynowak  wrote:

> This needs some explanation about the tradeoffs and what we recommend.

> For instance you should use session if you need to store a lot of data or if you rely on temp data for application correctness. If you are using sessions in general then you probably want to choose it for temp data.

Can you provide the complete text for explanation/tradeoffs? We can wordsmith.

**Comment by Eilon** [comment]
@Rick-Anderson I'll write something up and post here.

**Comment by Eilon** [comment]
Ok I updated the text. I added more details on choosing a provider. I also fixed some ordering/formatting of the headings because they seemed a bit off.

File: aspnetcore/fundamentals/app-state.md [hunk 3]
```diff
@@ -33,56 +33,70 @@ The server retains a session for a limited time after the last request. You can
...
+
+In ASP.NET Core 1.0 and 1.1, the session state TempData provider is the default.
+
+### Choosing a TempData provider
+
+Choosing a TempData provider involves several considerations, such as:
+
+1. Does the application use TempData only sparingly, for relatively small amounts of data (up to 500 bytes)? If so, the cookie TempData provider will add only a small cost to each request that carries TempData. If not, the session state TempData provider can be beneficial to avoid round-tripping a large amount of data in each request until the TempData is consumed.
```
Comment (rynowak) [inline aspnetcore/fundamentals/app-state.md]: > Does the application use TempData only sparingly

You can drop the `only` from this. 

I think that point 2 should come first, that's the most obvious consideration to me.

File: aspnetcore/fundamentals/app-state.md [hunk 4]
```diff
@@ -33,56 +33,70 @@ The server retains a session for a limited time after the last request. You can
...
 
-## Cookie-based TempData provider 
+### TempData providers
 
-In ASP.NET Core 1.1 and higher, you can use the cookie-based TempData provider to store a user's TempData in a cookie. To enable the  cookie-based TempData provider, register the `CookieTempDataProvider` service in `ConfigureServices`:
+In ASP.NET Core 2.0 and higher, the cookie-based TempData provider is used by default to store TempData in cookies.
+
+The cookie data is encoded with the [Base64UrlTextEncoder](/aspnet/core/api/microsoft.aspnetcore.authentication.base64urltextencoder). Because the cookie is encrypted and chunked, the single cookie size limit does not apply. The cookie data is not compressed because compressing encryped data can lead to security problems such as the [CRIME](https://wikipedia.org/wiki/CRIME_(security_exploit)) and [BREACH](https://wikipedia.org/wiki/BREACH_(security_exploit)) attacks. For more information on the cookie-based TempData provider, see [CookieTempDataProvider](https://github.com/aspnet/Mvc/blob/dev/src/Microsoft.AspNetCore.Mvc.ViewFeatures/ViewFeatures/CookieTempDataProvider.cs).
+
```
Comment (rynowak) [inline aspnetcore/fundamentals/app-state.md]: It would be nice to see it called out somewhere in plain English that **cookies have a size limit and thus cookie temp data has a size limit**

**Comment by Eilon** [comment]
@rynowak 🆙 📅

**Comment by rynowak** [comment]
You can update all you want, I already signed off.

**Comment by Eilon** [comment]
@Rick-Anderson if this looks good to you, can you squash & merge? Or else I'm happy to do so.
